### PR TITLE
How I Reduced My App's Network Usage by 95%

### PR DESCRIPTION
**Pull request by mcgill-a**
Please add a direct link to your post here:

https://mcgill-a.github.io/blog/2023/09/21/how-i-reduced-my-apps-network-usage-by-95-percent.html

Have you (please tick each box to show completion):

 - [x] Added your blog post to a single category?
 - [x] Added a brief summary for your post? Summaries should be roughly two sentences in length and give potential readers a good idea of the contents of your post.
 - [x] Checked that the build passes?
 - [x] Checked your spelling (you can use `npm spellcheck` if that's your thing)
 - [x] Ensured that your author profile contains a profile image, and a brief description of yourself? (make it more interesting than just your job title!)
 - [x] Optimised any images in your post? They should be less than 100KBytes as a general guide.

Posts are approved based on their category. replace `(at)` with `@` to notify them!

 - Tech - Colin Eberhardt - @ColinEberhardt
 - Cloud - Colin Eberhardt - (at)ColinEberhardt
 - Delivery - Matt Phillips - (at)scottlogicmphillips
 - UX - Graham Odds - (at)godds
 - People - Graham Odds - (at)godds
 - Testing - Laurence Pisani - (at)lpisani-SL

### CODE REVIEW

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+- Minimise state changes to reduce data transfers and re-renders.
+- Transmit only the necessary data changes to optimise updates, avoiding the transfer of entire objects.
+- Handle unstable internet connections by periodically querying for missed events.
+- Use an iterative development and testing cycle to resolve issues early and address user feedback.
+
+## Introduction
+
+In this blog post we'll take a brief look at some of the lessons learned while creating a cross-platform location sharing app.
+
+Why did I make this? Well, have a think - how many apps on your phone right now have the ability to share the real-time location of taxis, deliveries or people? If you had the same basic functionality you could make any number of services such as games, visualisations or even a rival delivery service. You'd be surprised how much fun some people can have with just the map itself ([looking at you GeoGuessr people](https://www.geoguessr.com)).
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: "Why did we make this _app_?" -  I was initially confused whether this question was regarding this blog or the app

Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: "how many apps on our phones have the ability to share...." might sound a bit nicer to me, not a big one!

Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: "You'd be surprised how much fun _users_ can have...." sounds  a bit nicer, and doesn't repeat "people" mentioned in the link text.

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+The app was built with Angular using [Ionic](https://ionicframework.com/) and [Capacitor](https://capacitorjs.com/). These tools give us the ability to build cross-platform apps using native features and view components using a single codebase. With any luck, it'll be released in the near future so here's a small subset of the core functionality:
+
+- Login with Google SSO to identify the user
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: "User authentication with Google SSO" maybe?

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+- Record user location data using background GPS tracking
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: Can we rephrase this to make it not sound like an app that tracks people without them knowing 😄 I might be tempted to remove this entry altogether and rephrase the one below as "Share user location data with other users" and "Match user location with pre-configured GPS landmarks server-side" or something like this?

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+## Findings from Test #1 with real users
+
+During a 1 hour test with 3 users here's what I found:
+
+- It used up a lot of battery
+- Data usage was much higher than expected (2GB per device on average)
+- Some of the state changes were completely missed (the app also broadcasts other important information)
+
+Clearly it needed some rethinking before the next test group. So what went wrong?
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: 👍 This section is great

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+One that could have been improved was the content of the data that was being sent. If one of the fields in the user object updated, the whole user object was sent. As it turns out, this is a huge waste of resources. This was exaggerated even more due the high volume of updates coming from the users.
+
+### 4. Unstable internet connections
+
+How often do you have a reliable connection on your phone when you're out and about? Will it stay connected if you go inside a building? It's not unlikely you'll be disconnected at some point - so what happens if you miss a WebSocket update?
+
+I didn't fancy over-engineering just yet so I settled on intermittently querying different parts of the backend to ask for various app states. More time sensitive data points were pinged more often. This was a great little trick to catch up on missed events but just to reiterate - make sure you carefully consider the frequency.
+
+## Findings from Test #2 with real users
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown:99]: Maybe a little bridge from the previous section to this section would be nice, not sure where it would fit will, but something along the lines of "With all of that fixed up and further minor improvements to the app, we then ran another test with real users..."

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+During a 2 hour test with 9 users here's what I found:
+
+- Battery usage was lower than expected
+- Data usage was much lower (using 100MB per device on average)
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: Much lower than test # 1? Or much lower than you expected?

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+How often do you have a reliable connection on your phone when you're out and about? Will it stay connected if you go inside a building? It's not unlikely you'll be disconnected at some point - so what happens if you miss a WebSocket update?
+
+I didn't fancy over-engineering just yet so I settled on intermittently querying different parts of the backend to ask for various app states. More time sensitive data points were pinged more often. This was a great little trick to catch up on missed events but just to reiterate - make sure you carefully consider the frequency.
+
+## Findings from Test #2 with real users
+
+During a 2 hour test with 9 users here's what I found:
+
+- Battery usage was lower than expected
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: lower than expected as compared to test # 1?

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+- Data usage was much lower (using 100MB per device on average)
+- If a user missed an update from the web socket, they received it shortly after
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: Bit cheesy but would it be wild to put the 👏 emojis after each of these bullet points? Take it or leave it :D

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown (hunk 7 above)
Comment (IMladjenovic) [reply _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: "Battery usage was lower than we feared" or "Battery usage was lower" if either of those apply? If its "Battery usage was lower than we had estimated" then some context on how you calculated this based on previous tests and improvements since then. Not sure on this one, you can leave this whole comment if you're happy with how it originally was

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown (hunk 6 above)
Comment (IMladjenovic) [reply _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: "Data usage was much lower than before" might make it a bit clearer?

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+## A comparison of the performance between Test #1 and Test #2
+
+![Data-Usage-Comparison]({{ site.github.url }}/amcgill/assets/data-usage-comparison.png "Data Usage Comparison")
+
+The second test lasted twice as long, had three times as many users, and we still managed to use 95% less data? Nice!
+
+Don't get me wrong - some bugs still cropped up, but that first test revealed a bunch of issues that I was able to address before the second test. This is why we test things in the real world with real users! If there's a bug, they'll probably find it.
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: "revealed many issues that I was able to address" maybe? I like this paragraph

File: _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+`---- show code snippet of original web sockets implementation ----`
+
+I had accidentally nested the socket listeners inside each active socket connection which meant the number of actions scales logarithmically proportional to the number of active users. Whoops! To make matters worse, some of these actions had side-effects which triggered another re-broadcast of the data, meaning those were sent out again too. This was a small bug but a huge oversight in my original approach that used far too much data.
+
+How did I miss this during development? Well, I:
+
+- Only tested it for a short amount of time
+- Was connected to a stable internet connection (Ethernet / WIFI)
+- Was plugged into a power outlet
```
Comment (IMladjenovic) [inline _posts/2023-10-02-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: These are cool insights

**Comment by ColinEberhardt** [comment]
Hi @mcgill-a thanks for submitting a blog post, looks like you're getting some detailed peer review from @IMladjenovic 

Once you're done with the review cycles, let me know and I'll do a final quick check ahead of publishing. Thanks :-)

**Comment by mcgill-a** [comment]
> Hi @mcgill-a thanks for submitting a blog post, looks like you're getting some detailed peer review from @IMladjenovic
> 
> Once you're done with the review cycles, let me know and I'll do a final quick check ahead of publishing. Thanks :-)

Thanks @ColinEberhardt I will do. I'll be posting it in the blog slack channel tomorrow to ask for more input and aim to have it published on October 2nd :)

File: _posts/2023-09-21-how-i-reduced-my-apps-network-usage-by-95-percent.markdown [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+Some of the key features are:
+
+- Authentication with Google SSO
+- Share location data with other users
+- Match location with pre-configured GPS landmarks server-side
+- Render avatars on a map and update user locations in real-time
+
+TODO: **add a gif of the location tracking here**
+
+I implemented these features this with a single codebase written in Angular, using [Ionic](https://ionicframework.com/) and [Capacitor](https://capacitorjs.com/) to build a cross-platform app with native features.
```
Comment (jstrong-scottlogic) [inline _posts/2023-09-21-how-i-reduced-my-apps-network-usage-by-95-percent.markdown]: 'these features this' -> 'these features'
